kauri/kauri.go

```diff
@@ -0,0 +1,301 @@
+	canMerge := true
+	if a == nil || b == nil {
+		k.logger.Info("one of it is nil")
+		return false
+	}
+	a.Participants().RangeWhile(func(i hotstuff.ID) bool {
+		b.Participants().RangeWhile(func(j hotstuff.ID) bool {
+			// cannot merge a and b if they both contain a contribution from the same ID.
+			if i == j {
+				canMerge = false
+			}
+			return canMerge
+		})
+		return canMerge
+	})
+	return canMerge
+}
+
+func (k *Kauri) verifyContribution(signature hotstuff.QuorumSignature, hash hotstuff.Hash) bool {
+	verified := false
+	block, ok := k.blockChain.Get(hash)
+	if !ok {
+		k.logger.Info("failed to fetch the block ", hash)
+		return verified
+	}
+	verified = k.crypto.Verify(signature, block.ToBytes())
+	return verified
+}
+
+func (k *Kauri) mergeWithContribution(currentSignature hotstuff.QuorumSignature) (bool, error) {
+	isVerified := k.verifyContribution(currentSignature, k.blockHash)
+	if !isVerified {
+		k.logger.Info("Contribution verification failed for view ", k.currentView,
+			"from participants", currentSignature.Participants(), " block hash ", k.blockHash)
+		return false, errors.New("unable to verify the contribution")
+	}
+	if k.aggregatedContribution == nil {
+		k.aggregatedContribution = currentSignature
+		return false, nil
+	}
+
+	if k.canMergeContributions(currentSignature, k.aggregatedContribution) {
+		new, err := k.crypto.Combine(currentSignature, k.aggregatedContribution)
+		if err == nil {
+			k.aggregatedContribution = new
+			if new.Participants().Len() >= k.configuration.QuorumSize() {
+				k.logger.Debug("Aggregated Complete QC and sending the event")
+				k.eventLoop.AddEvent(hotstuff.NewViewMsg{
+					SyncInfo: hotstuff.NewSyncInfo().WithQC(hotstuff.NewQuorumCert(
+						k.aggregatedContribution,
+						k.currentView,
+						k.blockHash,
+					)),
+				})
+				return true, nil
+			}
+		} else {
+			k.logger.Info("Failed to combine signatures: %v", err)
+			return false, errors.New("unable to combine signature")
+		}
+	} else {
+		k.logger.Debug("Failed to merge signatures due to overlap of signatures.")
+		return false, errors.New("unable to merge signature")
+	}
+	return false, nil
+}
+
+func isSubSet(a, b []hotstuff.ID) bool {
+	c := hotstuff.NewIDSet()
+	for _, id := range b {
+		c.Add(id)
+	}
+	for _, id := range a {
+		if !c.Contains(id) {
+			return false
+		}
+	}
+	return true
+}
+
+func (k *Kauri) randomizeIDS(hash hotstuff.Hash, leaderID hotstuff.ID) map[hotstuff.ID]int {
+	//assign leader to the root of the tree.
+
+	totalNodes := k.configuration.Len()
+	ids := make([]hotstuff.ID, 0, totalNodes)
+	for id := range k.configuration.Replicas() {
+		ids = append(ids, id)
+	}
+	sort.Slice(ids, func(i, j int) bool { return ids[i] < ids[j] })
+	seed := k.opts.SharedRandomSeed() + int64(binary.LittleEndian.Uint64(hash[:]))
+	//Shuffle the list of IDs using the shared random seed + the first 8 bytes of the hash.
+	rnd := rand.New(rand.NewSource(seed))
+	rnd.Shuffle(len(ids), reflect.Swapper(ids))
+	lIndex := 0
```

Comment:
Finding the leader index should use std lib func:
```go
lIndex := slices.Index(ids, leaderID)
if lIndex == -1 {
	panic("leader is not in the list of IDs")
}
```
The question is what to do if someone passes a leader that's not in the `ids` set.